rust/private/toolchain_utils.bzl

```diff
@@ -1,5 +1,171 @@
 """A module defining toolchain utilities"""
 
+def _symlink_sysroot_tree(ctx, name, target):
+    """Generate a set of symlinks to files from another target
+
+    Args:
+        ctx (ctx): The toolchain's context object
+        name (str): The name of the sysroot directory (typically `ctx.label.name`)
+        target (Target): A target owning files to symlink
+
+    Returns:
+        depset[File]: A depset of the generated symlink files
+    """
+    tree_files = []
+    for file in target.files.to_list():
```

Comment:
So this creates one symlink action per file in sysroot. How many files are there, and how large are they (extracted)?

Symlink action does a symlink (linux, mac, sometimes windows) or a copy (usually windows). On linux this is quite fast, on mac slower, on Windows very slow.

To be aware of the impact of this change, would it be an option to benchmark the impact of this change on cpu and memory use on a small project with full toolchain? I hope https://github.com/bazelbuild/bazel-bench works already :)

Reply:
Just to start dumping some data, This is what's produced for a compile action:
```
bazel-out/darwin-fastbuild/bin/external/rust_darwin_x86_64/toolchain_for_x86_64-apple-darwin_impl
├── bin
│   ├── cargo -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/bin/cargo
│   ├── clippy-driver -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/bin/clippy-driver
│   ├── rustc -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/bin/rustc
│   ├── rustdoc -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/bin/rustdoc
│   └── rustfmt -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/bin/rustfmt
└── lib
    ├── librustc-stable_rt.asan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/librustc-stable_rt.asan.dylib
    ├── librustc-stable_rt.lsan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/librustc-stable_rt.lsan.dylib
    ├── librustc-stable_rt.tsan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/librustc-stable_rt.tsan.dylib
    ├── librustc_driver-410e8e582906abf2.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/librustc_driver-410e8e582906abf2.dylib
    ├── libstd-91f0e69a7ebd6e00.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/libstd-91f0e69a7ebd6e00.dylib
    ├── libtest-cf5ce37eca3c5b4f.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/libtest-cf5ce37eca3c5b4f.dylib
    └── rustlib
        └── x86_64-apple-darwin
            ├── bin
            │   └── rust-lld -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/bin/rust-lld
            └── lib
                ├── libaddr2line-e4462cb73bc7c44d.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libaddr2line-e4462cb73bc7c44d.rlib
                ├── libadler-63585096303d8a9b.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libadler-63585096303d8a9b.rlib
                ├── liballoc-38142aee14781386.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/liballoc-38142aee14781386.rlib
                ├── libcfg_if-f9d3de4d6bdbf64b.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-f9d3de4d6bdbf64b.rlib
                ├── libcompiler_builtins-870fc7f82e4bde6f.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-870fc7f82e4bde6f.rlib
                ├── libcore-7eec28d9e637d9ae.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libcore-7eec28d9e637d9ae.rlib
                ├── libgetopts-c1e0b00ce77325b8.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libgetopts-c1e0b00ce77325b8.rlib
                ├── libgimli-37047837b923d0f0.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libgimli-37047837b923d0f0.rlib
                ├── libhashbrown-b493ccc9a0b06169.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-b493ccc9a0b06169.rlib
                ├── liblibc-b524a7407cf68766.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/liblibc-b524a7407cf68766.rlib
                ├── libmemchr-cadc9c82d273f933.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libmemchr-cadc9c82d273f933.rlib
                ├── libminiz_oxide-782208c813e9a91c.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libminiz_oxide-782208c813e9a91c.rlib
                ├── libobject-e31c8111c698d163.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libobject-e31c8111c698d163.rlib
                ├── libpanic_abort-e519a6588aa0d119.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libpanic_abort-e519a6588aa0d119.rlib
                ├── libpanic_unwind-36050f194975374d.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-36050f194975374d.rlib
                ├── libproc_macro-e654089d7f69e271.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libproc_macro-e654089d7f69e271.rlib
                ├── libprofiler_builtins-8ca8198cfac47c8f.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libprofiler_builtins-8ca8198cfac47c8f.rlib
                ├── librustc-stable_rt.asan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc-stable_rt.asan.dylib
                ├── librustc-stable_rt.lsan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc-stable_rt.lsan.dylib
                ├── librustc-stable_rt.tsan.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc-stable_rt.tsan.dylib
                ├── librustc_demangle-54ab9662306ce2de.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-54ab9662306ce2de.rlib
                ├── librustc_std_workspace_alloc-7a86533803c69aa4.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-7a86533803c69aa4.rlib
                ├── librustc_std_workspace_core-c27d31671763ac4a.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-c27d31671763ac4a.rlib
                ├── librustc_std_workspace_std-f7c6211f96295e24.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_std-f7c6211f96295e24.rlib
                ├── libstd-91f0e69a7ebd6e00.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libstd-91f0e69a7ebd6e00.dylib
                ├── libstd-91f0e69a7ebd6e00.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libstd-91f0e69a7ebd6e00.rlib
                ├── libstd_detect-16528d37a8f3eab3.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libstd_detect-16528d37a8f3eab3.rlib
                ├── libtest-cf5ce37eca3c5b4f.dylib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libtest-cf5ce37eca3c5b4f.dylib
                ├── libtest-cf5ce37eca3c5b4f.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libtest-cf5ce37eca3c5b4f.rlib
                ├── libunicode_width-5e001a9dfe8e0c3c.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libunicode_width-5e001a9dfe8e0c3c.rlib
                └── libunwind-95847f4481c1d402.rlib -> /private/var/tmp/_bazel_user/76282c66b0dfe3c5cb9a230bdc913a52/external/rust_darwin_x86_64/lib/rustlib/x86_64-apple-darwin/lib/libunwind-95847f4481c1d402.rlib
```

Reply:
> I hope https://github.com/bazelbuild/bazel-bench works already :)

I'm having issues getting it to run. Funny enough it's in the `pip install` step. The maintainers should use `rules_python` repository rules and make whatever this is doing just a Bazel command.

Reply:
@hlopko is there nothing available in Bazel CI that would show some metrics from the builds?

Reply:
Ok that looks acceptable, I don't worry anymore. Just to be extra sure, how much disk space do those files occupy?

Re bazel-bench, yeah it says WIP, but the equivalent script we have internally is super useful. If there is a concrete issue you're having feel free to file it on the repo, owners of it are all wonderful people, and I'm happy to ping them :)

Reply:
> Ok that looks acceptable, I don't worry anymore. Just to be extra sure, how much disk space do those files occupy?

```
rust_darwin_x86_64 % du -Lh -d0 toolchain_for_x86_64-apple-darwin_impl
386M	toolchain_for_x86_64-apple-darwin_impl
```

Seems reasonable(?)

> Re bazel-bench, yeah it says WIP, but the equivalent script we have internally is super useful. If there is a concrete issue you're having feel free to file it on the repo, owners of it are all wonderful people, and I'm happy to ping them :)

I definitely think it'd be worth pinging the group to do a sanity check and make sure it works outside of Google (don't know if it currently does or doesn't. I'm happy to find out the issue is only on my machine). That definitely seems like some useful functionality and would be great to run something like that on a regular basis 😄

Reply:
Yup I think that is reasonable (both things :))